src/cursor/cur_version.c

```diff
@@ -0,0 +1,432 @@
+            }
+
+            __wt_cursor_set_key(cursor, tw.start_txn, tw.start_ts, tw.durable_start_ts, stop_txn,
+              stop_ts, durable_stop_ts, 0, 0, 0, WT_VERSION_DISK_IMAGE);
+            cursor->value.data = vpack->data;
+            cursor->value.size = vpack->size;
+            F_SET(cursor, WT_CURSTD_VALUE_EXT);
+
+            version_cursor->upd_txnid = tw.start_txn;
+            version_cursor->upd_durable_stop_ts = tw.durable_start_ts;
+            version_cursor->upd_stop_ts = tw.start_ts;
+            version_cursor->hs_base_upd->data = vpack->data;
+            version_cursor->hs_base_upd->size = vpack->size;
+
+            upd_found = true;
+            F_SET(version_cursor, WT_VERSION_CUR_ON_DISK_EXHAUSTED);
+        }
+    }
+
+    if (!upd_found && !F_ISSET(version_cursor, WT_VERSION_CUR_HS_EXAUSTED)) {
+        /*
+         * If the history store cursor is not yet positioned, then we are traversing the history
+         * store versions for the first time.
+         */
+        if (!F_ISSET(hs_cursor, WT_CURSTD_KEY_INT)) {
+            hs_cursor->set_key(hs_cursor, 4, S2BT(session)->id, cursor->key, WT_TS_MAX, UINT64_MAX);
+            WT_ERR_NOTFOUND_OK(__wt_curhs_search_near_before(session, hs_cursor), true);
+        } else
+            WT_ERR_NOTFOUND_OK(ret = hs_cursor->prev(hs_cursor), true);
+
+        /*
+         * If there are no history store records for the given key or if we have iterated through
+         * all the records already, we have exhausted the history store.
+         */
+        if (ret == 0) {
+            WT_TIME_WINDOW_INIT(&tw);
+            hs_cursor->get_key(hs_cursor, &hs_btree_id, &hs_key, &tw.start_ts, &hs_counter);
+            hs_cursor->get_value(
+              hs_cursor, &tw.stop_ts, &tw.durable_start_ts, &hs_upd_type, &hs_value);
+            __wt_cursor_set_key(cursor, tw.start_txn, tw.start_ts, tw.durable_start_ts, tw.stop_txn,
+              tw.stop_ts, tw.durable_stop_ts, hs_upd_type, 0, 0, WT_VERSION_HISTORY_STORE);
+
+            /* Reconstruct the history store value if needed. */
+            if (hs_upd_type == WT_UPDATE_MODIFY) {
+                WT_ERR(__wt_modify_apply_item(
+                  session, table_cursor->value_format, version_cursor->hs_base_upd, hs_value.data));
+                cursor->value.data = version_cursor->hs_base_upd->data;
+                cursor->value.size = version_cursor->hs_base_upd->size;
+            } else {
+                WT_ASSERT(session, hs_upd_type == WT_UPDATE_STANDARD);
+                cursor->value.data = hs_value.data;
+                cursor->value.size = hs_value.size;
+                version_cursor->hs_base_upd->data = hs_value.data;
+                version_cursor->hs_base_upd->size = hs_value.size;
+            }
+            F_SET(cursor, WT_CURSTD_VALUE_EXT);
+            upd_found = true;
+        } else
+            F_SET(version_cursor, WT_VERSION_CUR_HS_EXAUSTED);
+    }
+
+done:
```

Comment:
This is wrong.

Reply:
done:
if (0) {
err:
        WT_TRET(cursor->reset(cursor));
    }
API_END_RET(session, ret);